File: Data Structures & Algorithms/swim-in-rising-water/submission-7.py
class Solution:
    def swimInWater(self, grid: List[List[int]]) -> int:
        rows = len(grid) 
        cols = len(grid[0])  
        visited = set()
        q = [(grid[0][0], 0, 0)]  
        dir = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        while q: 
            curH, r, c = heapq.heappop(q) 

            if r == rows - 1 and c == cols - 1: 
                return curH 

            for dr, dc in dir: 
                nr = r + dr 
                nc = c + dc 

                if 0 <= nr < rows and 0 <= nc < cols and (nr, nc) not in visited: 
                    visited.add((nr, nc)) 
                    heapq.heappush(q, (max(curH, grid[nr][nc]), nr, nc))

File: Data Structures & Algorithms/max-area-of-island/submission-7.js
class Solution {
    /**
     * @param {number[][]} grid
     * @return {number}
     */ 

    constructor() {
        this.dir = [[1, 0], [-1, 0], [0, 1], [0, -1]]
    }

    maxAreaOfIsland(grid) { 
        let visited = new Set()   
        let max = 0

        for (let r = 0; r < grid.length; r ++) {
            for (let c = 0; c < grid[0].length; c ++) { 
                if (grid[r][c] === 1 && !visited.has(`${r}_${c}`)) {
                    let area = this.dfs(grid, r, c, visited) 
                    max = Math.max(area, max)
                }
            }
        } 

        return max
     } 

    dfs(grid, r, c, visited) {
        if (!this.isValid(grid, r, c, visited)) {
            return 0
        } 
        
        visited.add(`${r}_${c}`)  
        let area = 1

        for (const [dr, dc] of this.dir) {
            const newR = r + dr 
            const newC = c + dc 
            area += this.dfs(grid, newR, newC, visited, area)
        } 

        return area
    } 

    isValid(grid, r, c, visited) {
        return ( 
            r >= 0 && 
            c >= 0 && 
            r < grid.length && 
            c < grid[0].length && 
            !visited.has(`${r}_${c}`) && 
            grid[r][c] === 1        
        )
    }
}
